src/nft/ZoraCreator1155Impl.sol

```diff
@@ -722,4 +723,67 @@ contract ZoraCreator1155Impl is
             revert();
         }
     }
+
+    /* start eip712 functionality */
+    bytes32 private constant _HASHED_NAME = keccak256(bytes("Preminter"));
+    bytes32 private constant _HASHED_VERSION = keccak256(bytes("1"));
+    bytes32 private constant _TYPE_HASH = keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
+
+    mapping(uint32 => bool) public uidUsed;
+
+    // todo: move to its own contract
+    error PremintAlreadyExecuted();
+    error MintNotYetStarted();
+    error PremintDeleted();
+
+    event CreatorAttribution(bytes32 structHash, bytes32 domainName, bytes32 version, bytes signature);
+
+    function delegateSetupNewToken(PremintConfig calldata premintConfig, bytes calldata signature) public returns (uint256 newTokenId) {
+        if (premintConfig.tokenConfig.mintStart != 0 && premintConfig.tokenConfig.mintStart > block.timestamp) {
+            // if the mint start is in the future, then revert
+            revert MintNotYetStarted();
+        }
+
+        // check that uid hasn't been used
+        if (uidUsed[premintConfig.uid]) {
+            revert PremintAlreadyExecuted();
+        } else {
+            uidUsed[premintConfig.uid] = true;
+        }
+
+        if (premintConfig.deleted) {
+            // if the signature says to be deleted, then dont execute any further minting logic;
+            // return 0
+            revert PremintDeleted();
+        }
+
+        bytes32 hashedPremintConfig = ZoraCreator1155Attribution.hashPremint(premintConfig);
+
+        // this is what attributes this token to have been created by the original creator
+        emit CreatorAttribution(hashedPremintConfig, ZoraCreator1155Attribution.HASHED_NAME, ZoraCreator1155Attribution.HASHED_VERSION, signature);
+
+        // recover the signer from the data
+        address recoveredSigner = ZoraCreator1155Attribution.recoverSignerHashed(hashedPremintConfig, signature, address(this), block.chainid);
+
+        // require that the signer can create new tokens (is a valid creator)
+        _requireAdminOrRole(recoveredSigner, CONTRACT_BASE_ID, PERMISSION_BIT_MINTER);
+
+        // temporarily grant msg sender admin permission to create new tokens
+        _addPermission(CONTRACT_BASE_ID, msg.sender, PERMISSION_BIT_MINTER);
```

Comment:
Why can't you just call `_setupNewToken` directly that skips the admin check passing in the user as the signer?